test/notifiex_test.exs

```diff
@@ -19,4 +19,32 @@ defmodule NotifiexTest do
 
     assert result == {:error, {:mock_error, false}}
   end
+
+  test "Slack notification must not work with missing params" do
+    expected_result = {:error, {:missing_required_params}, nil}
+
+    assert expected_result == Notifiex.send(:slack, nil, nil)
+    assert expected_result == Notifiex.send(:slack, nil, %{})
+    assert expected_result == Notifiex.send(:slack, %{}, nil)
+    assert expected_result == Notifiex.send(:slack, %{}, %{})
+
+    assert expected_result == Notifiex.send(:slack, %{text: "sample", channel: "channel"}, %{})
+    assert expected_result == Notifiex.send(:slack, %{text: "sample"}, %{token: "token"})
+    assert expected_result == Notifiex.send(:slack, %{channel: "channel"}, %{token: "token"})
+    assert expected_result == Notifiex.send(:slack, %{}, %{token: "token"})
+    assert expected_result == Notifiex.send(:slack, %{text: "sample"}, %{})
+    assert expected_result == Notifiex.send(:slack, %{channel: "channel"}, %{})
+  end
+
+  test "Discord notification must not work with missing params" do
+    expected_result = {:error, {:missing_required_params}, nil}
+
+    assert expected_result == Notifiex.send(:discord, nil, nil)
+    assert expected_result == Notifiex.send(:discord, nil, %{})
+    assert expected_result == Notifiex.send(:discord, %{}, nil)
+    assert expected_result == Notifiex.send(:discord, %{}, %{})
+
+    assert expected_result == Notifiex.send(:discord, %{content: "sample"}, %{})
+    assert expected_result == Notifiex.send(:discord, %{}, %{webhook: "webhook"})
```

Comment:
I like this, but I was wondering if we could rewrite this in a better way? If you think there isn't much room for improvement, please feel free to let me know!

I just feel that with more services being added, the number of checks would grow too.

Reply:
One way of writing that I come up is to use cartesian product. I have tested it and this works locally.
If you're okay with it, I'll push the commit

```elixir
test "Slack notification must not work with missing params", %{expected_missing_err: expected_missing_err} do
  for payload <- [nil, %{}, %{text: "sample", channel: "channel"}, %{text: "sample"}, %{channel: "channel"}],
      options <- [nil, %{}, %{token: "token"}],
      payload != %{text: "sample", channel: "channel"} && options != %{token: "token"}
      do
    assert expected_missing_err == Notifiex.send(:slack, payload, options)
  end
end

test "Discord notification must not work with missing params", %{expected_missing_err: expected_missing_err} do
  for payload <- [nil, %{}, %{content: "sample"}],
      options <- [nil, %{}, %{webhook: "webhook"}],
      payload != %{content: "sample"} && options != %{webhook: "webhook"}
      do
    assert expected_missing_err == Notifiex.send(:discord, payload, options)
  end
end
```

Reply:
This is what I was expecting, thanks! It would be nice if we could add a comment for the excluded case, which explains the reason behind it.